### DISCUSSION TITLE
Next.js App Router

### DISCUSSION THREAD

**Discussion by eugeniosegala** [reduxjs/redux-toolkit#3786] [General]
Hello! 

I'm using Redux RTK query on the server with Next.js, following the [official doc](https://redux-toolkit.js.org/rtk-query/usage/server-side-rendering).

I'm currently in the process of considering migrating my application to the ["App Router"](https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration) paradigm, however, I noticed that Redux RTK query is [not compatible](https://github.com/reduxjs/redux-toolkit/issues/3585) with that configuration at this time.

I started playing a bit with the new App Router feature and I noticed some limitations around accessing the full request object and other similar things. In general, it looks like what you can do from a server side prospective, it's more limited than before.

I was wondering, what is the position of the Redux team about this? It seems the App Router will become the standard at some point?

I'm happy to help figuring this out, perhaps we can reach out to the Next.js team and find a way to fix the above?

Thanks!

**Comment by markerikson** [comment]
Can you clarify the specific question you're asking?

Short answer is that Next and React are both looking to move more things to the server.  RTK Query is not as potentially relevant in that scenario.  But also there's a lot of technical complications towards potentially integrating RTKQ into a server component environment.  See https://phryneas.de/react-server-components-controversy for some details.

**Reply by eugeniosegala** [reply]
My question is:

How can I use the ["RTK Query Server Side Rendering feature"](https://redux-toolkit.js.org/rtk-query/usage/server-side-rendering) in combination with the Next JS ["App Router"](https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration) feature?

It can be used with the `getServerSideProps` [approach](https://github.com/phryneas/ssr-experiments/blob/main/nextjs-blog/pages/pokemon/ssr/%5Bname%5D.tsx) but it looks like it's not possible to use it with the new way Next JS decided to go forward.

I think there is still value in having RTK Query server side, mostly because you can:

- Be compliant with the ["Progressive enhancement"](https://en.wikipedia.org/wiki/Progressive_enhancement) pattern
- Fetch data [server-side](https://github.com/phryneas/ssr-experiments/blob/main/nextjs-blog/pages/pokemon/ssr/%5Bname%5D.tsx#L15) and rehydrate [client-side](https://github.com/phryneas/ssr-experiments/blob/main/nextjs-blog/pages/pokemon/%5Bname%5D.tsx#L12)
- Centralised source of truth for your data
- and more...

What would be the alternative at the moment to achieve similar results using Next.js App Router?

If there is appetite to make RTK work with the App Router, I'm happy to be involved.

I will read the article you shared :)

**Reply by markerikson** [reply]
We don't have a valid solution for that right now that I know of.  Based on Lenz's article, it seems that there's some technical pieces that simply don't exist on the React side.

We're certainly open and interested in potential solutions, but tbh right now all _my_ Redux attention is focused on shipping RTK 2.0.  Per my comment at https://github.com/reduxjs/redux-toolkit/issues/958#issuecomment-1742188595 , we're going to defer all major RTKQ changes for _after_ 2.0.

We'd definitely welcome help (both ideas and contributions), just setting expectations that this is not an immediate priority for us and I honestly don't know _if_ it's technically feasible.
